applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp

```diff
@@ -0,0 +1,603 @@
+        rData.U(i, TDim + 1) = r_geometry[i].FastGetSolutionStepValue(TOTAL_ENERGY);
+        rData.r(i) = r_geometry[i].FastGetSolutionStepValue(EXTERNAL_PRESSURE);
+    }
+
+    // Get shock capturing viscosity and heat conductivity
+    CalculateShockCapturingValues(rData);
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+double CompressibleNavierStokesExplicit<TDim, TNumNodes>::CalculateElementSize(BoundedMatrix<double,TNumNodes, TDim>& rDN_DX)
+{
+    double h = 0.0;
+    for (unsigned int i = 0; i < TNumNodes; ++i) {
+        double h_inv = 0.0;
+        for (unsigned int k = 0; k < TDim; ++k) {
+            h_inv += rDN_DX(i,k) * rDN_DX(i,k);
+        }
+        h += 1.0/h_inv;
+    }
+    h = sqrt(h) / static_cast<double>(TNumNodes);
+    return h;
+}
+
+template <>
+void CompressibleNavierStokesExplicit<2>::CalculateRightHandSide(
+    VectorType &rRightHandSideVector,
+    const ProcessInfo &rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    constexpr unsigned int n_nodes = 3;
+    constexpr unsigned int block_size = 4;
+    constexpr unsigned int matrix_size = n_nodes * block_size;
+
+    if (rRightHandSideVector.size() != matrix_size) {
+        rRightHandSideVector.resize(matrix_size, false); //false says not to preserve existing storage!!
+    }
+
+    // Struct to pass around the data
+    ElementDataStruct data;
+    this->FillElementData(data, rCurrentProcessInfo);
+
+    // Substitute the formulation symbols by the data structure values
+    const double h = data.h;
+    const array_1d<double, n_nodes> &r = data.r;
+    const BoundedMatrix<double, n_nodes, 2> &f_ext = data.f_ext;
+    const double mu = data.mu;
+    const double lambda = data.lambda;
+    const double c_v = data.c_v;
+    const double gamma = data.gamma;
+    const double v_sc = data.nu_sc;
+    const double k_sc = data.lambda_sc;
+
+    // Stabilization parameters
+    const double stab_c1 = 4.0;
+    const double stab_c2 = 2.0;
+
+    // Solution vector values from nodal data
+    // This is intentionally done in this way to limit the matrix acceses
+    // The notation U_i_j DOF j value in node i
+    const double &U_0_0 = data.U(0, 0);
+    const double &U_0_1 = data.U(0, 1);
+    const double &U_0_2 = data.U(0, 2);
+    const double &U_0_3 = data.U(0, 3);
+    const double &U_1_0 = data.U(1, 0);
+    const double &U_1_1 = data.U(1, 1);
+    const double &U_1_2 = data.U(1, 2);
+    const double &U_1_3 = data.U(1, 3);
+    const double &U_2_0 = data.U(2, 0);
+    const double &U_2_1 = data.U(2, 1);
+    const double &U_2_2 = data.U(2, 2);
+    const double &U_2_3 = data.U(2, 3);
+
+    // Hardcoded shape functions for linear triangular element
+    // This is explicitly done to minimize the allocation and matrix acceses
+    // The notation N_i_j means shape function for node j in Gauss pt. i
+    const double one_sixt = 1.0/6.0;
+    const double two_third = 2.0/3.0;
+    const double N_0_0 = one_sixt;

```

Comment:
in a future PR i would put this in the symbolic so to enjoy simplifications.

Reply:
Done.